beep.cpp

```diff
@@ -0,0 +1,227 @@
+#include "beep.h"
+
+Beep::Beep(QObject *parent) :
+    QObject(parent)
+{
+    m_wavePtr = NULL;
+    m_open = false;
+}
+
+Beep::~Beep()
+{
+    deinit();
+    if (m_wavePtr) delete(m_wavePtr);
+}
+
+bool Beep::init()
+{
+    if (isOpen()) {
+        qDebug() << "{QML] sound card is already open";
+        return true;
+    }
+
+    // Open audio card we wish to use for playback
+    register int err;
+    if ((err = snd_pcm_open(&m_playbackHandle, &SoundCardPortName[0], SND_PCM_STREAM_PLAYBACK, 0)) < 0)
+    {
+        qDebug("[QML] can't open audio %s: %s\n", &SoundCardPortName[0], snd_strerror(err));
+        return false;
+    }
+    else
+      qDebug() << "[QML] sound card is open";
+
+    m_open = true;
+    return true;
+
+}
+
+void Beep::deinit()
+{
+    if (isOpen())
+    {
+        snd_pcm_close(m_playbackHandle);
+        m_open = false;
+        qDebug() << "[QML] sound card closed";
+    }
+}
+
+bool Beep::isOpen()
+{
+    return m_open;
+}
+
+bool Beep::openwave(const QString &path)
+{
+    // No wave data loaded yet
+    m_wavePtr = NULL;
+
+    if (!loadWaveFile(path.toUtf8().data())) {
+        return false;
+    }
+    else {
+        // Open audio card we wish to use for playback
+        if (isOpen()) {
+
+            switch (m_waveBits)
+            {
+                case 8:
+                    m_format = SND_PCM_FORMAT_U8;
+                    break;
+
+                case 16:
+                    m_format = SND_PCM_FORMAT_S16;
+                    break;
+
+                case 24:
+                    m_format = SND_PCM_FORMAT_S24;
+                    break;
+
+                case 32:
+                    m_format = SND_PCM_FORMAT_S32;
+                    break;
+            }
+
+        }
+    }
+
+    return true;
+}
+
+void Beep::play()
+{
+    if (m_wavePtr && isOpen())
+    {
+        register snd_pcm_uframes_t	count, frames;
+        register int err;
+        // Set the audio card's hardware parameters (sample rate, bit resolution, etc)
+        if ((err = snd_pcm_set_params(m_playbackHandle, m_format, SND_PCM_ACCESS_RW_INTERLEAVED, m_waveChannels, m_waveRate, 1, 100000)) < 0)
```

Comment:
Can snd_pcm_set_params() be in init()? None of the params are changing between calls to play. Then in here call snd_pcm_prepare() before the output loop.